### ISSUE TITLE
DPO with local fine-tuned llama 3.1 70b not progressing

### ISSUE THREAD

**Issue by enqidu** [pytorch/torchtune#2130]
this is config file:

`
# Model Arguments
model:
  _component_: torchtune.models.llama3_1.lora_llama3_1_70b
  lora_attn_modules: ['q_proj', 'v_proj', 'output_proj']
  apply_lora_to_mlp: False
  apply_lora_to_output: False
  lora_rank: 8  # higher increases accuracy and memory
  lora_alpha: 16  # usually alpha=2*rank
  lora_dropout: 0.0

tokenizer:
  _component_: torchtune.models.llama3.llama3_tokenizer
  path: ./dachi_new_model_70b_3.1-instruct/tokenizer.model

checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: ./dachi_new_model_70b_3.1-instruct
  checkpoint_files: [
    model-00001-of-00030.safetensors,
    model-00002-of-00030.safetensors,
    model-00003-of-00030.safetensors,
    model-00004-of-00030.safetensors,
    model-00005-of-00030.safetensors,
    model-00006-of-00030.safetensors,
    model-00007-of-00030.safetensors,
    model-00008-of-00030.safetensors,
    model-00009-of-00030.safetensors,
    model-00010-of-00030.safetensors,
    model-00011-of-00030.safetensors,
    model-00012-of-00030.safetensors,
    model-00013-of-00030.safetensors,
    model-00014-of-00030.safetensors,
    model-00015-of-00030.safetensors,
    model-00016-of-00030.safetensors,
    model-00017-of-00030.safetensors,
    model-00018-of-00030.safetensors,
    model-00019-of-00030.safetensors,
    model-00020-of-00030.safetensors,
    model-00021-of-00030.safetensors,
    model-00022-of-00030.safetensors,
    model-00023-of-00030.safetensors,
    model-00024-of-00030.safetensors,
    model-00025-of-00030.safetensors,
    model-00026-of-00030.safetensors,
    model-00027-of-00030.safetensors,
    model-00028-of-00030.safetensors,
    model-00029-of-00030.safetensors,
    model-00030-of-00030.safetensors,
  ]
  recipe_checkpoint: null
  output_dir: ./tmp/Meta-Llama-3.1-70B-instruct-dpo
  model_type: LLAMA3
resume_from_checkpoint: False
save_adapter_weights_only: False

# Dataset and Sampler
dataset:
  _component_: torchtune.datasets.preference_dataset
  source: csv
  data_files: Combined/DPO/combined_dpo_examples.csv
  column_map:
    chosen: Output_e
    rejected: Output_b
  split: train
seed: null
shuffle: True
batch_size: 1
n_accumulate: 4

# Optimizer and Scheduler
optimizer:
  _component_: torch.optim.AdamW
  fused: True
  weight_decay: 0.05
  lr: 5e-4
lr_scheduler:
  _component_: torchtune.training.lr_schedulers.get_cosine_schedule_with_warmup
  num_warmup_steps: 100

loss:
  _component_: torchtune.rlhf.loss.DPOLoss
  beta: 0.1
  label_smoothing: 0

# Training
epochs: 1
max_steps_per_epoch: 1000
gradient_accumulation_steps: 8  # Use to increase virtual batch size
compile: False  # pytorch compile, set to true for better perf/memory

# Logging
output_dir: ./tmp/lora_dpo_output/
metric_logger:
  _component_: torchtune.training.metric_logging.DiskLogger
  log_dir: ${output_dir}
log_every_n_steps: 1
log_peak_memory_stats: True

# Environment
device: cuda
dtype: bf16
enable_activation_checkpointing: True  # True reduces memory
enable_activation_offloading: False  # True reduces memory`

Output is frozen on the message: 

`INFO:torchtune.utils._logging:FSDP is enabled. Instantiating model and loading checkpoint on Rank 0 ...
`

**Comment by SalmanMohammadi** [comment]
Hey @enqidu! Thanks for checking torchtune out : ))

I'd like to try reproduce the behaviour you're seeing here. Could you share your hardware configuration, and the command which you used to run the recipe?

Aside, I noticed:
```yaml
n_accumulate: 4
```
In your config. Is this a custom recipe which you're finetuning with? I'm curious what you're using this parameter for : )

**Comment by enqidu** [comment]
Hello @SalmanMohammadi  

I'm running on ml.p4de.24xlarge instance - 8 NVIDIA A100

using lora_dpo_distributed recipe.

Regarding the 

`n_accumulate: 4`

Most of the config is copied from llama 3_1 8b DPO config, I haven't explicitly checked each parameter.

**Comment by SalmanMohammadi** [comment]
Hey @enqidu. I tried reproducing this using your config on a 8xA100 node but wasn't able to. I'm using the nightly version of torchtune and the latest stable installation of torch
```bash
root@bc943fa118d5:~/torchtune# tune run --nnodes 1 --nproc_per_node 8 lora_dpo_distributed --config /workspace/config.yaml 
Running with torchrun...
W1209 18:31:53.618000 16723 torch/distributed/run.py:793] 
W1209 18:31:53.618000 16723 torch/distributed/run.py:793] *****************************************
W1209 18:31:53.618000 16723 torch/distributed/run.py:793] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
W1209 18:31:53.618000 16723 torch/distributed/run.py:793] *****************************************
INFO:torchtune.utils._logging:Running LoRADPORecipeDistributed with resolved config:

batch_size: 1
checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: /workspace/Meta-Llama-3.1-70B-Instruct/
  checkpoint_files:
  - model-00001-of-00030.safetensors
  - model-00002-of-00030.safetensors
  - model-00003-of-00030.safetensors
  - model-00004-of-00030.safetensors
  - model-00005-of-00030.safetensors
  - model-00006-of-00030.safetensors
  - model-00007-of-00030.safetensors
  - model-00008-of-00030.safetensors
  - model-00009-of-00030.safetensors
  - model-00010-of-00030.safetensors
  - model-00011-of-00030.safetensors
  - model-00012-of-00030.safetensors
  - model-00013-of-00030.safetensors
  - model-00014-of-00030.safetensors
  - model-00015-of-00030.safetensors
  - model-00016-of-00030.safetensors
  - model-00017-of-00030.safetensors
  - model-00018-of-00030.safetensors
  - model-00019-of-00030.safetensors
  - model-00020-of-00030.safetensors
  - model-00021-of-00030.safetensors
  - model-00022-of-00030.safetensors
  - model-00023-of-00030.safetensors
  - model-00024-of-00030.safetensors
  - model-00025-of-00030.safetensors
  - model-00026-of-00030.safetensors
  - model-00027-of-00030.safetensors
  - model-00028-of-00030.safetensors
  - model-00029-of-00030.safetensors
  - model-00030-of-00030.safetensors
  model_type: LLAMA3
  output_dir: /workspace/tmp/Meta-Llama-3.1-70B-Instruct/
  recipe_checkpoint: null
compile: false
dataset:
  _component_: torchtune.datasets.stack_exchange_paired_dataset
device: cuda
dtype: bf16
enable_activation_checkpointing: true
enable_activation_offloading: false
epochs: 1
gradient_accumulation_steps: 8
log_every_n_steps: 1
log_peak_memory_stats: true
loss:
  _component_: torchtune.rlhf.loss.DPOLoss
  beta: 0.1
  label_smoothing: 0
lr_scheduler:
  _component_: torchtune.training.lr_schedulers.get_cosine_schedule_with_warmup
  num_warmup_steps: 100
max_steps_per_epoch: 1000
metric_logger:
  _component_: torchtune.training.metric_logging.StdoutLogger
  log_dir: /workspace/tmp/lora_dpo_output/
model:
  _component_: torchtune.models.llama3_1.lora_llama3_1_70b
  apply_lora_to_mlp: false
  apply_lora_to_output: false
  lora_alpha: 16
  lora_attn_modules:
  - q_proj
  - v_proj
  - output_proj
  lora_dropout: 0.0
  lora_rank: 8
n_accumulate: 4
optimizer:
  _component_: torch.optim.AdamW
  fused: true
  lr: 0.0005
  weight_decay: 0.05
output_dir: /workspace/tmp/lora_dpo_output/
resume_from_checkpoint: false
save_adapter_weights_only: true
seed: null
shuffle: true
tokenizer:
  _component_: torchtune.models.llama3.llama3_tokenizer
  path: /workspace/Meta-Llama-3.1-70B-Instruct/original/tokenizer.model

INFO:torchtune.utils._logging:Hint: enable_activation_checkpointing is True, but enable_activation_offloading isn't. Enabling activation offloading should reduce memory further.
DEBUG:torchtune.utils._logging:Setting manual seed to local seed 940935333. Local seed is seed + rank = 940935333 + 0
INFO:torchtune.utils._logging:metric logger is initialized.
INFO:torchtune.utils._logging:FSDP is enabled. Instantiating model and loading checkpoint on Rank 0 ...
INFO:torchtune.utils._logging:Instantiating model and loading checkpoint took 55.68 secs
```

Would you be able to share which versions of torchtune and torch you're using?
cc @ebsmothers if there's something obvious here

**Comment by enqidu** [comment]
dataset:
  _component_: torchtune.datasets.preference_dataset
  source: csv
  data_files: Combined/DPO/combined_dpo_examples.csv
  column_map:
    chosen: Output_e
    rejected: Output_b
  split: train

Might it be a dataset problem?

**Comment by SalmanMohammadi** [comment]
> `INFO:torchtune.utils._logging:FSDP is enabled. Instantiating model and loading checkpoint on Rank 0 ...`

If you're hanging here the issue is in trying to instantiate the model, which happens before we setup data. 
As a sanity check, could you try running the normal LoRA distributed recipe using [this config](https://github.com/pytorch/torchtune/blob/main/recipes/configs/llama3_1/70B_lora.yaml) - modifying the checkpoint and tokenizer directories to point to where you've downloaded your models?

If this also fails it may be something due to your setup.

**Comment by enqidu** [comment]
Wow, seems like I'm having a similar problem with LoRA distributed recipe. Definitely, something is wrong with my setup but what could be the reason?

Thanks very much

**Comment by yechenzhi** [comment]
try 
```
pip uninstall torchao 
pip install torchao
pip install -e .["dev"]
```
I'm not sure why, but in my case it worked.

**Comment by yechenzhi** [comment]
try `export NCCL_P2P_DISABLE=1`, I double there is  no p2p in your gpus.

**Comment by enqidu** [comment]
Thanks @yechenzhi !
Solving the problem has progressed so far.
Things I changed, I updated to new Torchtune version and fine-tuned llama-3.1-70b again.
Ran with TORCH_CPP_LOG_LEVEL=INFO NCCL_DEBUG=INFO  

So far DPO started working, and the model is initialized as far as I can see.

Now I have a problem with the dataset, but this problem is solved for now.

**Comment by enqidu** [comment]
OK, so I will write about another problem later but as I can see

DPO trainer doesn't save adapter_config.json file after training. I think it might cause some problems for users.

**Comment by enqidu** [comment]
Regarding the main problem about output being frozen on the message:

`INFO:torchtune.utils._logging:FSDP is enabled. Instantiating model and loading checkpoint on Rank 0 ...
`

As I can see, it only progresses if checkpoints are already cached. It has trouble reading from the disk and the problem persists.

**Comment by yechenzhi** [comment]
@SalmanMohammadi @enqidu I'm not sure. Previously, I encountered the same issue and debugged it, finding it stuck [here](https://github.com/pytorch/pytorch/blob/297ce776363cc4802fa74d210fced2b4128960d5/torch/distributed/distributed_c10d.py#L4110), I suspect the scatter operation requires P2P communication, but since the GPUs lack P2P support, it waits indefinitely. Setting export NCCL_P2P_DISABLE=1 allows PyTorch to use alternative communication methods between GPUs, though it significantly slows down performance. Disabling P2P enables the code to run, but at a much slower speed.

**Comment by SalmanMohammadi** [comment]
@yechenzhi I also experienced this behaviour on 2xRTX6000s. I found a couple related discussions (e.g. https://discuss.pytorch.org/t/ddp-training-on-rtx-4090-ada-cu118/168366, https://forums.developer.nvidia.com/t/standard-nvidia-cuda-tests-fail-with-dual-rtx-4090-linux-box/233202/59) - the issue seemed to be around P2P not being enabled for certain cards (Ada desktop cards)/bugs in older NVIDIA drivers. It's weird that @enqidu was seeing this on A800s, maybe something unique to their hardware setup. 

I think disabling P2P communication or trying to update your drivers by following the discussion in the NVIDIA dev forum linked above might be one way to get around this. I'm going to resolve this issue for now. Thanks so much for your help @yechenzhi!

**Comment by enqidu** [comment]
I just want to reiterate that export NCCL_P2P_DISABLE=1 didn't work for me. It was (and still is) a model initialization issue with checkpoint shards. I found a workaround solution to read the model and then have torchtune read it from cache.

**Comment by SalmanMohammadi** [comment]
> I just want to address that `export NCCL_P2P_DISABLE=1 ` didn't work for me. It was (and still is) a model initialization issue with checkpoint shards. I found a workaround solution to read the model and then have torchtune read it from cache.

Sorry, I misunderstood your earlier comment! Could you share a bit more here? What was your workaround solution?

**Comment by yechenzhi** [comment]
@enqidu Could you run` nvidia-smi topo -m` and share the output with us?
